gateway/ln-gateway/src/gateway_module_v2/complete_sm.rs

```diff
@@ -95,49 +94,37 @@ impl CompleteStateMachine {
     async fn await_receive(
         context: GatewayClientContextV2,
         operation_id: OperationId,
-    ) -> Result<[u8; 32], String> {
-        let mut stream = context.notifier.subscribe(operation_id).await;
-
-        loop {
-            if let Some(GatewayClientStateMachinesV2::Receive(state)) = stream.next().await {
-                match state.state {
-                    ReceiveSMState::Funding => {}
-                    ReceiveSMState::Rejected(error) => {
-                        return Err(error);
-                    }
-                    ReceiveSMState::Success(preimage) => {
-                        return Ok(preimage);
-                    }
-                    ReceiveSMState::Failure => {
-                        return Err("Failure".to_string());
-                    }
-                    ReceiveSMState::Refunding(..) => {
-                        return Err("Refunding".to_string());
-                    }
-                }
-            }
-        }
+    ) -> FinalReceiveState {
+        context.module.await_receive(operation_id).await
     }
 
     fn transition_receive(
-        result: Result<[u8; 32], String>,
+        final_receive_state: FinalReceiveState,
         old_state: &CompleteStateMachine,
     ) -> CompleteStateMachine {
-        old_state.update(CompleteSMState::Completing(result))
+        old_state.update(CompleteSMState::Completing(final_receive_state))
     }
 
     async fn await_completion(
         context: GatewayClientContextV2,
         payment_hash: bitcoin_hashes::sha256::Hash,
+        final_receive_state: FinalReceiveState,
         incoming_chan_id: u64,
         htlc_id: u64,
-        result: Result<[u8; 32], String>,
     ) {
-        let action = match result {
-            Ok(preimage) => Action::Settle(Settle {
+        let action = match final_receive_state {
+            FinalReceiveState::Rejected => Action::Cancel(Cancel {
+                reason: "Rejected".to_string(),
```

Comment:
These error reasons will be incredibly hard to debug if they ever happen in production

Reply:
If I understand correctly these error will be returned to the sender of the lightning payment correct? The full error is saved in the state machine in the gateway so that the operator could look it up - does it make sense to return the error to the sender of the lightning payment?

Reply:
Let's iterate if there's issues in the future